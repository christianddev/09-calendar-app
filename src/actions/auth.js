import Swal from "sweetalert2";
import { customFetch, customFetchToken } from "../helpers/fetch";
import { processErrorMessage } from "../helpers/processErrorMessage";
import { types } from "../types/types";

const setToken = (token) => {
  localStorage.setItem("token", token);
  localStorage.setItem("token-init-date", new Date().getTime());
};

const login = (user) => ({
  type: types.authLogin,
  payload: user,
});

const register = (user) => ({
  type: types.authStartRegister,
  payload: user,
});

const checkingFinish = () => ({
  type: types.authCheckingFinish,
});

const logout = () => ({ type: types.authLogout });

export const startLogin = (email, password) => {
  return async (dispatch) => {
    const response = await customFetch("auth", { email, password }, "POST");
    const body = await response.json();
    if (body.ok) {
      setToken(body.token);
      dispatch(
        login({
          uid: body.uid,
          name: body.name,
        })
      );
    } else {
      Swal.fire("Error", processErrorMessage(body?.errors), "error");
    }
  };
};

export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    const response = await customFetch(
      "auth/new",
      { name, email, password },
      "POST"
    );
    const body = await response.json();
    if (body.ok) {
      setToken(body.token);
      dispatch(
        register({
          uid: body.uid,
          name: body.name,
        })
      );
    } else {
      Swal.fire("Error", processErrorMessage(body?.errors), "error");
    }
  };
};

export const startChecking = () => {
  return async (dispatch) => {
    const response = await customFetchToken("auth/renew");
    const body = await response.json();
    if (body.ok) {
      setToken(body.token);
      dispatch(
        login({
          uid: body.uid,
          name: body.name,
        })
      );
    } else {
      dispatch(checkingFinish());
    }
  };
};

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(logout());
  };
};
