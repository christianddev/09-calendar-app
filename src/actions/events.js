import Swal from "sweetalert2";
import { customFetchToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { processErrorMessage } from "../helpers/processErrorMessage";
import { types } from "../types/types";

const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event,
});

export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {
    try {
      const {
        auth: { uid: _id, name },
      } = getState();
      const response = await customFetchToken("events", event, "POST");
      const body = await response.json();
      if (body.ok) {
        event.id = body.msg.event.id;
        event.user = {
          _id,
          name,
        };
        dispatch(eventAddNew(event));
      } else {
        Swal.fire("Error", processErrorMessage(body?.errors), "error");
      }
    } catch (error) {
      console.error("eventStartAddNew", error);
      Swal.fire("Error", JSON.stringify(error), "error");
    }
  };
};

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event,
});

export const eventClearActiveEvent = () => ({
  type: types.eventClearActiveEvent,
});

export const eventStartUpdated = (event) => {
  return async (dispatch) => {
    try {
      const response = await customFetchToken(
        `events/${event.id}`,
        event,
        "PUT"
      );
      const body = await response.json();
      if (body.ok) {
        dispatch(eventUpdated(event));
      } else {
        Swal.fire("Error", processErrorMessage(body?.errors), "error");
      }
    } catch (error) {
      console.error("eventStartUpdated", error);
    }
  };
};

const eventUpdated = (event) => ({
  type: types.eventUpdated,
  payload: event,
});

export const eventStartDelete = () => {
  return async (dispatch, getState) => {
    try {
      const { id } = getState()?.calendar?.activeEvent;
      const response = await customFetchToken(`events/${id}`, {}, "DELETE");
      const body = await response.json();
      if (body.ok) {
        dispatch(eventDeleted());
      } else {
        Swal.fire("Error", processErrorMessage(body?.errors), "error");
      }
    } catch (error) {
      console.error("eventStartUpdated", error);
    }
  };
};

const eventDeleted = () => ({
  type: types.eventDeleted,
});

export const eventStartLoading = () => {
  return async (dispatch) => {
    try {
      const response = await customFetchToken("events");
      const body = await response.json();
      const events = prepareEvents(body?.msg?.events);
      dispatch(eventLoaded(events));
    } catch (error) {
      console.error("eventStartLoading", error);
    }
  };
};

const eventLoaded = (events) => ({
  type: types.eventLoaded,
  payload: events,
});

export const eventLogout = () => ({ type: types.eventLogout });
