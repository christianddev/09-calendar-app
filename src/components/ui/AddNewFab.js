import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { uiOpenModal } from '../../actions/ui'

export const AddNewFab = () => {
  const dispatch = useDispatch()

  const { activeEvent } = useSelector((state) => state.calendar);
  const handleClick = () =>  dispatch(uiOpenModal())
  return (
    <button
      className='btn btn-primary fab'
      onClick={handleClick}
    >
      {activeEvent
        ? <i className='fas fa-edit'></i>
        : <i className='fas fa-plus'></i>
      }
    </button>
  )
}
