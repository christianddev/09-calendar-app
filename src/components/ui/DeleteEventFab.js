import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { eventDeleted } from "../../actions/events";

export const DeleteEventFab = () => {
  const dispatch = useDispatch();
  const { activeEvent } = useSelector((state) => state.calendar);

  const handleClick = () => dispatch(eventDeleted());
  return (
    <button
      className="btn btn-danger fab-danger"
      onClick={handleClick}
      disabled={!activeEvent}
    >
      <i className="fas fa-trash"></i>
    </button>
  );
};
