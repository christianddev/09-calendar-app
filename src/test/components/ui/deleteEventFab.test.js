import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import "@testing-library/jest-dom";
import { MemoryRouter } from "react-router-dom";

import { DeleteEventFab } from "../../../components/ui/DeleteEventFab";
import { eventStartDelete } from "../../../actions/events";

jest.mock("../../../actions/events", () => ({
  eventStartDelete: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
  calendar: {
    activeEvent: {
      title: "",
      notes: "",
      start: "",
      end: "",
      user: {
        _id: "",
        name: "",
      },
      id: "",
    },
  },
};
const store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <DeleteEventFab />
    </MemoryRouter>
  </Provider>
);

describe("deleteEventFab.js", () => {
  test("should show correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should call eventStartDelete", () => {
    wrapper.find(".btn-danger").prop("onClick")();
    expect(eventStartDelete).toHaveBeenCalledTimes(1);
  });
});
