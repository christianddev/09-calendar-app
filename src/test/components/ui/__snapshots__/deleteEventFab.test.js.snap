// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deleteEventFab.js should show correctly 1`] = `
<button
  className="btn btn-danger fab-danger"
  disabled={false}
  onClick={[Function]}
>
  <i
    className="fas fa-trash"
  />
</button>
`;
